File: WPF/MVVM/DelegateCommand.vb
Imports System.Windows.Input

Namespace Common.MVVM

    Public Class DelegateCommand
        Implements ICommand

        Public Sub New(ByVal ExecuteFunc As Action)
            Me._ExecuteFunc = Sub(O) ExecuteFunc.Invoke()
            Me._CanExecuteFunc = Nothing
        End Sub

        Public Sub New(ByVal ExecuteFunc As Action, ByVal CanExecuteFunc As Func(Of Boolean))
            Me._ExecuteFunc = Sub(O) ExecuteFunc.Invoke()
            Me._CanExecuteFunc = Function(O) CanExecuteFunc.Invoke()
        End Sub

        Public Sub New(ByVal ExecuteFunc As Action(Of Object))
            Me._ExecuteFunc = ExecuteFunc
            Me._CanExecuteFunc = Nothing
        End Sub

        Public Sub New(ByVal ExecuteFunc As Action(Of Object), ByVal CanExecuteFunc As Func(Of Object, Boolean))
            Me._ExecuteFunc = ExecuteFunc
            Me._CanExecuteFunc = CanExecuteFunc
        End Sub

        Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            Me._ExecuteFunc.Invoke(parameter)
        End Sub

        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            Return If(Me._CanExecuteFunc Is Nothing, True, Me._CanExecuteFunc.Invoke(parameter))
        End Function

        Private ReadOnly _ExecuteFunc As Action(Of Object)
        Private ReadOnly _CanExecuteFunc As Func(Of Object, Boolean)

    End Class

End Namespace

File: WPF/MVVM/DelegateCommand.cs
using System;
using System.Windows.Input;

namespace Ks
{
    namespace Ks.Common.MVVM
    {
        public class DelegateCommand : ICommand
        {
            public DelegateCommand(Action ExecuteFunc)
            {
                this._ExecuteFunc = O => ExecuteFunc.Invoke();
                this._CanExecuteFunc = null;
            }

            public DelegateCommand(Action ExecuteFunc, Func<bool> CanExecuteFunc)
            {
                this._ExecuteFunc = O => ExecuteFunc.Invoke();
                this._CanExecuteFunc = O => CanExecuteFunc.Invoke();
            }

            public DelegateCommand(Action<object> ExecuteFunc)
            {
                this._ExecuteFunc = ExecuteFunc;
                this._CanExecuteFunc = null;
            }

            public DelegateCommand(Action<object> ExecuteFunc, Func<object, bool> CanExecuteFunc)
            {
                this._ExecuteFunc = ExecuteFunc;
                this._CanExecuteFunc = CanExecuteFunc;
            }

            public event EventHandler CanExecuteChanged;

            public void Execute(object parameter)
            {
                this._ExecuteFunc.Invoke(parameter);
            }

            public bool CanExecute(object parameter)
            {
                return this._CanExecuteFunc?.Invoke(parameter) ?? true;
            }

            private readonly Action<object> _ExecuteFunc;
            private readonly Func<object, bool> _CanExecuteFunc;
        }
    }
}
